autogen/oai/client.py

```diff
@@ -107,17 +312,17 @@ def __init__(self, *, config_list: Optional[List[Dict[str, Any]]] = None, **base
         openai_config, extra_kwargs = self._separate_openai_config(base_config)
         if type(config_list) is list and len(config_list) == 0:
             logger.warning("openai client was provided with an empty config_list, which may not be intended.")
+        self._clients: List[Client] = []
         if config_list:
             config_list = [config.copy() for config in config_list]  # make a copy before modifying
-            self._clients: List[OpenAI] = [
-                self._client(config, openai_config) for config in config_list
-            ]  # could modify the config
+            for config in config_list:
+                self._register_client(config, openai_config)  # could modify the config
```

Comment:
I think we can be more explict here:

```python
for config in config_list:
  if "api_type" not in config or config["api_type"] is None:
    self._register_default_client(config, openai_config)
  elif config["api_type"] == "openai":
    self.register_client("openai", OpenAIClient, **config, **openai_config)
  elif config["api_type"] == "azure_openai":
    # registers azure.
  elif config["api_type"] == "gemini":
    # register gemini 
  # ... more can be added here in the future as we add more support. 
```

Perhaps a better idea is to have a dictionary that maps api type to corresponding client class. Easier for us to accept more contributions. 

We can expose that global dictionary so users can modify and override as they wish in their application code.

Reply:
If we replace config dictionaries with Pydantic classes, then we can have this functionality implemented there. It would make this code one-liner and much easier to extend. That should be done in a separate PR anyway.

Reply:
ideally we reach a point where adding another client doesn't need to be hardcoded in the code